Add JSON Test Vectors

siv2r opened the pull request:
Addresses issue #35.

This PR introduces functionality for generating and validating JSON test vectors for ChillDKG’s public APIs.

Summary:

- Implements [Design 3](https://github.com/BlockstreamResearch/bip-frost-dkg/issues/35#issuecomment-2565855499), with assertions for [intermediate messages](https://github.com/BlockstreamResearch/bip-frost-dkg/issues/35#issuecomment-2577725803).
- Modules in `python/vectors_generator/` generate one JSON vector file per API, saving them in the `python/vectors/`.
- Modifies `chilldkg_ref`:
    1. To ensure reproducible test vectors, this PR temporarily uses a zeroed `aux_rand` value for `schnorr_sign`
    2. Adds the `GE.from_bytes_with_infinity` method.
- Adds tests within `tests.py` to load the generated JSON vectors and verify API outputs for correctness.

siv2r committed:
python: add test vectors

outputs one JSON vector file per API in `python/vectors` directory

jonasnick reviewed (commented):
Nice, thank you! This looks like "design 3" as I had it in mind.

I was surprised, shouldn't this be based on #88? I think the test vectors should contain the messages in their serialized form.

jonasnick commented on python/vector_generator/coordinator.py:
```diff
@@ -0,0 +1,247 @@
+            "cmsg1": cmsg1_asdict(cmsg1),
+            "pmsgs2": [pmsg2_asdict(m) for m in invalid_pmsgs2],
+            "error": error,
+            "comment": "invalid certificate length",
```
wrong comment

jonasnick commented on python/vector_generator/participant.py:
```diff
@@ -0,0 +1,532 @@
+            "pmsg1": pmsg1_asdict(pmsgs1[0]),
+            "cmsg1": cmsg1_asdict(invalid_cmsg1),
+            "error": error,
+            "comment": "coms_to_secret list in cmsg1 has an invalid value at index 0",
```
wrong comment

jonasnick commented on python/vector_generator/participant.py:
```diff
@@ -0,0 +1,532 @@
+            "pmsg1": pmsg1_asdict(pmsgs1[0]),
+            "cmsg1": cmsg1_asdict(invalid_cmsg1),
+            "error": error,
+            "comment": "coms_to_secret list in cmsg1 has an invalid value at index 0",
```
wrong comment

jonasnick commented on python/vector_generator/others.py:
```diff
@@ -0,0 +1,352 @@
+def generate_hostpubkey_vectors():
+    vectors = {"valid_test_cases": [], "error_test_cases": []}
+
+    # --- Valid test case 1 ---
```
nit: can we start numbering with 0? The test cases are stored in a list, where the conventional first index is 0.

jonasnick reviewed (commented):
Posting a status update, still reviewing...

siv2r commented:
Thanks for the review. I'll address your comments soon. 

> I was surprised, shouldn't this be based on https://github.com/BlockstreamResearch/bip-frost-dkg/pull/88? I think the test vectors should contain the messages in their serialized form.

I assumed we were favoring design 2 over design 1. That is why the JSON representation of messages shows internal fields (for better readability) rather than a single byte array. The definitions of design 1 and design 2 are in [this comment](https://github.com/BlockstreamResearch/bip-frost-dkg/issues/35#issuecomment-2553866102) as:

> We have two main designs for our custom types in the JSON:
> 
> 1. **Design 1**: Represent it as a single large hex string (a fully serialized byte array).
> 2. **Design 2**: Use a nested JSON structure that shows all internal fields.

I can switch the message JSON to a single hex-encoded byte array if you prefer.

siv2r commented on python/vector_generator/others.py:
```diff
@@ -0,0 +1,352 @@
+    # --- Error test case 3: last share in enc_secshare list is out of range ---
+    n = len(hostpubkeys)
+    cert_len = chilldkg.certeq_cert_len(n)
+    rand_encshare = bytes.fromhex(
+        "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141"
+    )
```
Ah yes, thanks.

jonasnick commented:
I had thought design 1 vs. design 2 was mostly about representing the state objects. As for the messages that are passed between the participants, I believe it's better to include them in the test vectors in their serialized form. This way, the byte encoding is tested as well, which is an important part of the spec.

jonasnick commented on python/tests.py:
```diff
@@ -362,6 +381,306 @@ def test_correctness(t, n, simulate_dkg, recovery=False, investigation=False):
 
+def test_hostpubkey_gen_vectors():
+    input_file = Path("vectors/hostpubkey_gen_vectors.json")
+    with open(input_file) as f:
```
When run from the parent directory via `python python/tests.py` this gives me an error:
```
FileNotFoundError: [Errno 2] No such file or directory: 'vectors/hostpubkey_gen_vectors.json'
```
`python python/tests.py` works fine on master

siv2r replied:
Fixed

jonasnick commented on python/vector_generator/coordinator.py:
```diff
@@ -0,0 +1,247 @@
+    invalid_pmsgs1[1].enc_pmsg.enc_shares.pop()
+    error = expect_exception(
+        lambda: coordinator_step1(invalid_pmsgs1, params),
+        chilldkg.FaultyParticipantOrCoordinatorError,  # REVIEW: why is this working? shouldn't it raise an error saying encpedpop.FaultyParticipantOrCoordinatorError was raised instead?
```
Why would it raise "`encpedpop.FaultyParticipantOrCoordinatorError`"? `FaultyParticipantOrCoordinatorError` is defined in util.py and it's part of the chilldkg module.

siv2r replied:
Yes, makes sense. Thank you.

mllwchrry reviewed (commented):
I've identified several coverage gaps in the test vectors, aside from those mentioned by @jonasnick and the TODO statements. Consider adding cases below.

General improvements:
- Add valid test cases for different threshold settings (not just 2-of-3).
- Many fields are currently only tested at index 0 or 1; ensure logic is tested against both the caller's index and some other participant index. This is especially important for operations that validate or process data per-participant.
- Whenever `hostseckey` is used, along with existing tests, test for zero key and `hostseckey` >= group order.

Participants logic:
- In `participant_step1` t > n test is missing.
- In `participant_finalize` there's a test for an invalid certificate length with too short certificate; consider adding a check where the certificate is too long (same for `coordinator_finalize`).

Recovery logic:
- A test for duplicate `pubnonces` in the `pubnonces` list is missing.

jonasnick commented on python/gen_vector_utils/coordinator.py:
```diff
@@ -0,0 +1,247 @@
+import chilldkg_ref.chilldkg as chilldkg
+
+
+def generate_coordinator_step1_vectors():
```
Let's defer more exhaustive test coverage to subsequent PRs.

siv2r commented on python/gen_vector_utils/participant.py:
```diff
@@ -0,0 +1,532 @@
+        }
+    )
+
+    # --- Error Test Case 1: Wrong host secret key length ---
```
It should be have been removed. Fixed.

siv2r commented on python/gen_vector_utils/coordinator.py:
```diff
@@ -0,0 +1,247 @@
+            "pmsgs1": [pmsg1_asdict(m) for m in pmsgs1],
+            "params": params_asdict(invalid_params),
+            "error": error,
+            "comment": "hostpubkeys list contains and invalid value",
```
fixed

siv2r commented on python/gen_vector_utils/participant.py:
```diff
@@ -0,0 +1,532 @@
+            "E5CFC54DA8EE57BA97C389060D00BB840A9DDF6BF1E32AE3D3598373EF384EE7",
+        ]
+    )
+    assert len(randoms) == len(hostpubkeys)
```
Ah yes, I kept this check in case these two lists are modified incorrectly in the future. They need to stay coupled. I think it's safe to keep it?

siv2r commented:
@jonasnick @mllwchrry, this PR is ready for another round of review!

Since the last review, a couple of things have changed upstream (message ser/deser merged, `aux_rand` parameter added to `participant_step2`). On top of that, here's what's new:
- Participant and coordinator messages are now encoded as serialized bytes in the JSON vectors, instead of the earlier nested-value format.
- Test vectors are fully reproducible, `aux_rand` is set to a fixed value so re-running `gen_vectors.py` always produces the same output.

I've addressed most of jonasnick's review comments:
- Reordered the error test cases to match the validation order in the code (e.g., threshold is checked before duplicate hostpubkeys).
- Added a test for wrong randomness length in `participant_step1`.
- Added a test for tampered `sum_coms_to_nonconst_terms` in `participant_step2`.
- Fixed `FileNotFoundError` when running `python python/tests.py` from the repo root.
- Various other small fixes (comments, naming, documentation).

I also brought in a few ideas from the wycheproof test vector format: specifically `tc_id`, `description`, and `total_tests` fields. The bigger wycheproof patterns (test groups referencing shared fields, valid/invalid/acceptable tags) weren't needed since we already have shared fields at the top of each JSON file and separate lists for valid and error test cases.

These vectors aren't exhaustive. This PR establishes the structure and baseline for how ChillDKG test vectors will look. More thorough coverage will come in follow-up PRs.

siv2r commented:
One more thing, regarding testing different `(t, n)` configurations (like mllwchrry mentioned): we could adopt wycheproof's `testGroups` pattern for this. Each group would carry its own `params`, shared setup fields, and test cases. That way, we can test multiple configurations (e.g., standard `n=3, t=2` alongside degenerate `n=1, t=1`) in a single file while keeping `tc_id` unique across groups. Something to consider for a follow-up PR.

```json                                                                                                                                                      
{                                                                                                                                                            
  "testGroups": [                                                                                                                                            
    {                                                                                                                                                        
      "description": "Standard 2-of-3 DKG",                                                                                                                  
      "params": { "hostpubkeys": [...], "t": 2 },                                                                                                            
      "hostseckey": "ADE1...",                                                                                                                               
      "random": "42B5...",                                                                                                                                   
      "aux_rand": "005F...",                                                                                                                                 
      "pmsg1": "036B...",                                                                                                                                    
      "valid_test_cases": [...],                                                                                                                             
      "error_test_cases": [...]                                                                                                                              
    },                                                                                                                                                       
    {                                                                                                                                                        
      "description": "Degenerate 1-of-1 DKG",                                                                                                                
      "params": { "hostpubkeys": ["03AB..."], "t": 1 },                                                                                                      
      "hostseckey": "...",                                                                                                                                   
      "random": "...",                                                                                                                                       
      "aux_rand": "...",                                                                                                                                     
      "pmsg1": "...",                                                                                                                                        
      "valid_test_cases": [...],                                                                                                                             
      "error_test_cases": [...]                                                                                                                              
    }                                                                                                                                                        
  ]                                                                                                                                                          
}                                                                                                                                                            
```

siv2r commented:
Our pool arrays (`pmsg1_pool`, `cmsg1_pool`, etc.) are just lists of hex blobs with no metadata. When a test case references `pmsg1_indices: [0, 1, 3]`, you have to guess what pool entry 3 is.

We could add a parallel `_annotations` array next to each pool, one string per entry describing what it is and, for invalid entries, what was changed:

```json
{
  "pmsg1_pool": [
    "036BEE122E...",
    "037C42E200...",
    "026F714517...",
    "037C42E200...truncated"
  ],
  "pmsg1_pool_annotations": [
    "valid pmsg1 from participant 0",
    "valid pmsg1 from participant 1",
    "valid pmsg1 from participant 2",
    "participant 1's pmsg1 truncated by 32 bytes (missing encrypted shares)."
  ]
}
```

The generator script already knows what it's constructing for each entry, so this is basically free to maintain. Would like to know if you think it's worth including this extra `_annotations` field.

mllwchrry commented:
The PR is good now. Found only a few typos and one suggestion:

1. Typo in `participant.py`. Lines 270, 286, 289, 303 have comments referring to `coms_to_secret` but the actual field name is `coms_to_secrets`.
2. The same test data (`hostseckeys`, `randoms`, `aux_rand`, invalid signatures, etc.) is duplicated across different files for generating vectors. These could be extracted into a separate file with constants, which would make maintenance easier. However, this could be done in another PR.

I also like the ideas you proposed for `testGroups` and `_annotations`. They are great enhancements!

siv2r commented on python/gen_vector_utils/participant.py:
```diff
@@ -0,0 +1,532 @@
+            "comment": "host secret key doesn't match any hostpubkey",
+        }
+    )
+    # --- Error test case 3: Invalid threshold ---
```
Fixed

siv2r commented on python/gen_vector_utils/participant.py:
```diff
@@ -0,0 +1,532 @@
+    return vectors
+
+
+def generate_participant_step2_vectors():
```
Done

jonasnick commented:
The code looks good and implements the design we discussed. Let's defer solving the remaining issues to follow-up PRs. Thanks @siv2r and @mllwchrry!
